usr/src/man/man1/chgrp.1
.\"
.\" Sun Microsystems, Inc. gratefully acknowledges The Open Group for
.\" permission to reproduce portions of its copyrighted documentation.
.\" Original documentation from The Open Group can be obtained online at
.\" http://www.opengroup.org/bookstore/.
.\"
.\" The Institute of Electrical and Electronics Engineers and The Open
.\" Group, have given us permission to reprint portions of their
.\" documentation.
.\"
.\" In the following statement, the phrase ``this text'' refers to portions
.\" of the system documentation.
.\"
.\" Portions of this text are reprinted and reproduced in electronic form
.\" in the SunOS Reference Manual, from IEEE Std 1003.1, 2004 Edition,
.\" Standard for Information Technology -- Portable Operating System
.\" Interface (POSIX), The Open Group Base Specifications Issue 6,
.\" Copyright (C) 2001-2004 by the Institute of Electrical and Electronics
.\" Engineers, Inc and The Open Group.  In the event of any discrepancy
.\" between these versions and the original IEEE and The Open Group
.\" Standard, the original IEEE and The Open Group Standard is the referee
.\" document.  The original Standard can be obtained online at
.\" http://www.opengroup.org/unix/online.html.
.\"
.\" This notice shall appear on any product containing this material.
.\"
.\" The Berkeley software License Agreement specifies the terms and conditions
.\" for redistribution.
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" or http://www.opensolaris.org/os/licensing.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\"
.\" Copyright (c) 1983 Regents of the University of California.
.\" All rights reserved.
.\" Copyright 1989 AT&T
.\" Portions Copyright (c) 1992, X/Open Company Limited  All Rights Reserved
.\" Copyright (c) 2003, Sun Microsystems, Inc.  All Rights Reserved
.\"
.TH CHGRP 1 "Feb 21, 2019"
.SH NAME
chgrp \- change file group ownership
.SH SYNOPSIS
.LP
.nf
\fBchgrp\fR [\fB-fhR\fR] \fIgroup\fR \fIfile\fR...
.fi

.LP
.nf
\fBchgrp\fR \fB-s\fR [\fB-fhR\fR] \fIgroupsid\fR \fIfile\fR...
.fi

.LP
.nf
\fBchgrp\fR \fB-R\fR [f] [\fB-H\fR | \fB-L\fR | \fB-P\fR] \fIgroup\fR \fIfile\fR...
.fi

.LP
.nf
\fBchgrp\fR \fB-s\fR \fB-R\fR [f] [\fB-H\fR | \fB-L\fR | \fB-P\fR] \fIgroupsid\fR \fIfile\fR...
.fi

.SH DESCRIPTION
.LP
The \fBchgrp\fR utility will set the group ID of the file named by each
\fIfile\fR operand to the group ID specified by the \fIgroup\fR operand.
.sp
.LP
For each \fIfile\fR operand, it will perform actions equivalent to the
\fBchown\fR(2) function, called with the following arguments:
.RS +4
.TP
.ie t \(bu
.el o
The \fIfile\fR operand will be used as the \fIpath\fR argument.
.RE
.RS +4
.TP
.ie t \(bu
.el o
The user ID of the file will be used as the \fIowner\fR argument.
.RE
.RS +4
.TP
.ie t \(bu
.el o
The specified group ID will be used as the \fIgroup\fR argument.
.RE
.sp
.LP
Unless \fBchgrp\fR is invoked by a process with appropriate privileges, the
set-user-ID and set-group-ID bits of a regular file will be cleared upon
successful completion; the set-user-ID and set-group-ID bits of other file
types may be cleared.
.sp
.LP
The operating system has a configuration option \fB_POSIX_CHOWN_RESTRICTED\fR,
to restrict ownership changes. When this option is in effect, the owner of the
file may change the group of the file only to a group to which the owner
belongs. Only the super-user can arbitrarily change owner IDs, whether or not
this option is in effect. To set this configuration option, include the
following line in \fB/etc/system\fR:
.sp
.in +2
.nf
set rstchown = 1
.fi
.in -2

.sp
.LP
To disable this option, include the following line in \fB/etc/system\fR:
.sp
.in +2
.nf
set rstchown = 0
.fi
.in -2

.sp
.LP
\fB_POSIX_CHOWN_RESTRICTED\fR is enabled by default. See \fBsystem\fR(5) and
\fBfpathconf\fR(2).
.SH OPTIONS
.LP
The following options are supported.

.sp
.ne 2
.na
\fB\fB-f\fR\fR
.ad
.RS 6n
Force. Does not report errors.
.RE

.sp
.ne 2
.na
\fB\fB-h\fR\fR
.ad
.RS 6n
If the file is a symbolic link, this option changes the group of the symbolic
link. Without this option, the group of the file referenced by the symbolic
link is changed.
.RE

.sp
.ne 2
.na
\fB\fB-H\fR\fR
.ad
.RS 6n
If the file specified on the command line is a symbolic link referencing a file
of type directory, this option changes the group of the directory referenced by
the symbolic link and all the files in the file hierarchy below it. If a
symbolic link is encountered when traversing a file hierarchy, the group of the
target file is changed, but no recursion takes place.
.RE

.sp
.ne 2
.na
\fB\fB-L\fR\fR
.ad
.RS 6n
If the file is a symbolic link, this option changes the group of the file
referenced by the symbolic link. If the file specified on the command line, or
encountered during the traversal of the file hierarchy, is a symbolic link
referencing a file of type directory, then this option changes the group of the
directory referenced by the symbolic link and all files in the file hierarchy
below it.
.RE

.sp
.ne 2
.na
\fB\fB-P\fR\fR
.ad
.RS 6n
If the file specified on the command line or encountered during the traversal
of a file hierarchy is a symbolic link, this option changes the group of the
symbolic link. This option does not follow the symbolic link to any other part
of the file hierarchy.
.RE

.sp
.ne 2
.na
\fB\fB-R\fR\fR
.ad
.RS 6n
Recursive. \fBchgrp\fR descends through the directory, and any subdirectories,
setting the specified group \fBID\fR as it proceeds. When a symbolic link is
encountered, the group of the of the symbolic link is changed, unless the
\fB-H\fR or \fB-L\fR option is specified. Unless the \fB-H\fR, \fB-L\fR,
or \fB-P\fR option is specified, the \fB-P\fR option is used as the default mode.
.RE

.sp
.ne 2
.na
\fB\fB-s\fR\fR
.ad
.RS 6n
The specified group is Windows SID. This option requires a file system that
supports storing SIDs, such as ZFS.
.RE

.sp
.LP
Specifying more than one of the mutually-exclusive options \fB-H\fR, \fB-L\fR,
or \fB-P\fR is not considered an error. The last option specified determines
the behavior of \fBchgrp\fR.

.SH OPERANDS
.LP
The following operands are supported:
.sp
.ne 2
.na
\fB\fIgroup\fR\fR
.ad
.RS 9n
A group name from the group database or a numeric group ID. Either specifies a
group ID to be given to each file named by one of the \fIfile\fR operands. If a
numeric \fIgroup\fR operand exists in the group database as a group name, the
group ID number associated with that group name is used as the group ID.
.RE

.sp
.ne 2
.na
\fB\fIfile\fR\fR
.ad
.RS 9n
A path name of a file whose group ID is to be modified.
.RE

.SH USAGE
.LP
See \fBlargefile\fR(7) for the description of the behavior of \fBchgrp\fR when
encountering files greater than or equal to 2 Gbyte (2^31 bytes).
.SH ENVIRONMENT VARIABLES
.LP
See \fBenviron\fR(7) for descriptions of the following environment variables
that affect the execution of \fBchgrp\fR: \fBLANG\fR, \fBLC_ALL\fR,
\fBLC_CTYPE\fR, \fBLC_MESSAGES\fR, and \fBNLSPATH\fR.
.SH EXIT STATUS
.LP
The following exit values are returned:
.sp
.ne 2
.na
\fB\fB0\fR\fR
.ad
.RS 6n
The utility executed successfully and all requested changes were made.
.RE

.sp
.ne 2
.na
\fB\fB>0\fR\fR
.ad
.RS 6n
An error occurred.
.RE

.SH FILES
.ne 2
.na
\fB\fB/etc/group\fR\fR
.ad
.RS 14n
group file
.RE

.SH ATTRIBUTES
.LP
See \fBattributes\fR(7) for descriptions of the following attributes:

.TS
box;
c | c
l | l .
ATTRIBUTE TYPE	ATTRIBUTE VALUE
_
CSI	Enabled. See NOTES.
_
Interface Stability	Committed
_
Standard	See \fBstandards\fR(7).
.TE

.SH SEE ALSO
.LP
.BR chmod (1),
.BR chown (1),
.BR chown (2),
.BR fpathconf (2),
.BR group (5),
.BR passwd (5),
.BR system (5),
.BR attributes (7),
.BR environ (7),
.BR largefile (7),
.BR standards (7),
.BR id (8)
.SH NOTES
.LP
\fBchgrp\fR is CSI-enabled except for the \fIgroup\fR name.
.sp
.LP
In the past the behavior of \fB/usr/xpg4/bin/chgrp\fR and
\fB/usr/bin/chgrp\fR utilities was different. Now they behave the same way.


usr/src/contrib/ast/src/lib/libcmd/chgrp.c
/***********************************************************************
*                                                                      *
*               This software is part of the ast package               *
*          Copyright (c) 1992-2012 AT&T Intellectual Property          *
*                      and is licensed under the                       *
*                 Eclipse Public License, Version 1.0                  *
*                    by AT&T Intellectual Property                     *
*                                                                      *
*                A copy of the License is available at                 *
*          http://www.eclipse.org/org/documents/epl-v10.html           *
*         (with md5 checksum b35adb5213ca9657e911e9befb180842)         *
*                                                                      *
*              Information and Software Systems Research               *
*                            AT&T Research                             *
*                           Florham Park NJ                            *
*                                                                      *
*                 Glenn Fowler <gsf@research.att.com>                  *
*                  David Korn <dgk@research.att.com>                   *
*                                                                      *
***********************************************************************/
#pragma prototyped
/*
 * David Korn
 * Glenn Fowler
 * AT&T Research
 *
 * chgrp+chown
 */

static const char usage_1[] =
"[-?@(#)$Id: chgrp (AT&T Research) 2012-04-20 $\n]"
USAGE_LICENSE
;

static const char usage_grp_1[] =
"[+NAME?chgrp - change the group ownership of files]"
"[+DESCRIPTION?\bchgrp\b changes the group ownership of each file"
"	to \agroup\a, which can be either a group name or a numeric"
"	group id. The user ownership of each file may also be changed to"
"	\auser\a by prepending \auser\a\b:\b to the group name.]"
;

static const char usage_own_1[] =
"[+NAME?chown - change the ownership of files]"
"[+DESCRIPTION?\bchown\b changes the ownership of each file"
"	to \auser\a, which can be either a user name or a numeric"
"	user id. The group ownership of each file may also be changed to"
"	\auser\a by appending \b:\b\agroup\a to the user name.]"
;

static const char usage_2[] =
"[b:before?Only change files with \bctime\b before (less than) the "
    "\bmtime\b of \afile\a.]:[file]"
"[c:changes?Describe only files whose ownership actually changes.]"
"[f:quiet|silent?Do not report files whose ownership fails to change.]"
"[h|l:symlink?Change the ownership of symbolic links on systems that "
    "support \blchown\b(2). Implies \b--physical\b.]"
"[m:map?The first operand is interpreted as a file that contains a map "
    "of space separated \afrom_uid:from_gid to_uid:to_gid\a pairs. The "
    "\auid\a or \agid\a part of each pair may be omitted to mean any \auid\a "
    "or \agid\a. Ownership of files matching the \afrom\a part of any pair "
    "is changed to the corresponding \ato\a part of the pair. The matching "
    "for each file operand is in the order \auid\a:\agid\a, \auid\a:, "
    ":\agid\a. For a given file, once a \auid\a or \agid\a mapping is "
    "determined it is not overridden by any subsequent match. Unmatched "
    "files are silently ignored.]"
"[n:show?Show actions but don't execute.]"
"[N:numeric?By default numeric user and group id operands are first "
    "interpreted as names; if no name exists then they are interpreted as "
    "explicit numeric ids. \b--numeric\b interprets numeric id operands as "
    "numeric ids.]"
"[r:reference?Omit the explicit ownership operand and use the ownership "
    "of \afile\a instead.]:[file]"
"[u:unmapped?Print a diagnostic for each file for which either the "
    "\auid\a or \agid\a or both were not mapped.]"
"[v:verbose?Describe changed permissions of all files.]"
"[H:metaphysical?Follow symbolic links for command arguments; otherwise "
    "don't follow symbolic links when traversing directories.]"
"[L:logical|follow?Follow symbolic links when traversing directories.]"
"[P:physical|nofollow?Don't follow symbolic links when traversing "
    "directories.]"
"[R:recursive?Recursively change ownership of directories and their "
    "contents.]"
"[X:test?Canonicalize output for testing.]"

"\n"
"\n"
;

static const char usage_3[] =
" file ...\n"
"\n"
"[+EXIT STATUS?]{"
	"[+0?All files changed successfully.]"
	"[+>0?Unable to change ownership of one or more files.]"
"}"
"[+SEE ALSO?\bchmod\b(1), \bchown\b(2), \btw\b(1), \bgetconf\b(1), \bls\b(1)]"
;

#if defined(__STDPP__directive) && defined(__STDPP__hide)
__STDPP__directive pragma pp:hide lchown
#else
#define lchown		______lchown
#endif

#include <cmd.h>
#include <cdt.h>
#include <ls.h>
#include <ctype.h>
#include <fts_fix.h>

#ifndef ENOSYS
#define ENOSYS	EINVAL
#endif

#include "FEATURE/symlink"

#if defined(__STDPP__directive) && defined(__STDPP__hide)
__STDPP__directive pragma pp:nohide lchown
#else
#undef	lchown
#endif

typedef struct Key_s			/* uid/gid key			*/
{
	int		uid;		/* uid				*/
	int		gid;		/* gid				*/
} Key_t;

typedef struct Map_s			/* uid/gid map			*/
{
	Dtlink_t	link;		/* dictionary link		*/
	Key_t		key;		/* key				*/
	Key_t		to;		/* map to these			*/
} Map_t;

/*
 * libast's struid() has the peculiar feature that it returns -1 in response
 * to not finding a UID for a name on the first call, and -2 for subsequent
 * calls for the same string.
 */
#define NOID		(-1)
#define STILLNOID	(-2)

#define OPT_CHOWN	0x0001		/* chown			*/
#define OPT_FORCE	0x0002		/* ignore errors		*/
#define OPT_GID		0x0004		/* have gid			*/
#define OPT_LCHOWN	0x0008		/* lchown			*/
#define OPT_NUMERIC	0x0010		/* favor numeric ids		*/
#define OPT_SHOW	0x0020		/* show but don't do		*/
#define OPT_TEST	0x0040		/* canonicalize output		*/
#define OPT_UID		0x0080		/* have uid			*/
#define OPT_UNMAPPED	0x0100		/* unmapped file diagnostic	*/
#define OPT_VERBOSE	0x0200		/* have uid			*/

extern int	lchown(const char*, uid_t, gid_t);

/*
 * parse uid and gid from s
 */

static void
getids(register char* s, char** e, Key_t* key, int options)
{
	register char*	t;
	register int	n;
	register int	m;
	char*		z;
	char		buf[64];

	key->uid = key->gid = NOID;
	while (isspace(*s))
		s++;
	for (t = s; (n = *t) && n != ':' && n != '.' && !isspace(n); t++);
	if (n)
	{
		options |= OPT_CHOWN;
		if ((n = t++ - s) >= sizeof(buf))
			n = sizeof(buf) - 1;
		*((s = (char*)memcpy(buf, s, n)) + n) = 0;
	}
	if (options & OPT_CHOWN)
	{
		if (*s)
		{
			n = (int)strtol(s, &z, 0);
			if (*z || !(options & OPT_NUMERIC))
			{
				if ((m = struid(s)) != NOID && m != STILLNOID)
					n = m;
				else if (*z)
					error(ERROR_exit(1), "%s: unknown user", s);
			}
			key->uid = n;
		}
		for (s = t; (n = *t) && !isspace(n); t++);
		if (n)
		{
			if ((n = t++ - s) >= sizeof(buf))
				n = sizeof(buf) - 1;
			*((s = (char*)memcpy(buf, s, n)) + n) = 0;
		}
	}
	if (*s)
	{
		n = (int)strtol(s, &z, 0);
		if (*z || !(options & OPT_NUMERIC))
		{
			if ((m = strgid(s)) != NOID && m != STILLNOID)
				n = m;
			else if (*z)
				error(ERROR_exit(1), "%s: unknown group", s);
		}
		key->gid = n;
	}
	if (e)
		*e = t;
}

/*
 * NOTE: we only use the native lchown() on symlinks just in case
 *	 the implementation is a feckless stub
 */

int
b_chgrp(int argc, char** argv, Shbltin_t* context)
{
	register int	options = 0;
	register char*	s;
	register Map_t*	m;
	register FTS*	fts;
	register FTSENT*ent;
	register int	i;
	Dt_t*		map = 0;
	int		logical = 1;
	int		flags;
	int		uid;
	int		gid;
	char*		op;
	char*		usage;
	char*		t;
	Sfio_t*		sp;
	unsigned long	before;
	Dtdisc_t	mapdisc;
	Key_t		keys[3];
	Key_t		key;
	struct stat	st;
	int		(*chownf)(const char*, uid_t, gid_t);

	cmdinit(argc, argv, context, ERROR_CATALOG, ERROR_NOTIFY);
	flags = fts_flags() | FTS_META | FTS_TOP | FTS_NOPOSTORDER | FTS_NOSEEDOTDIR;
	before = ~0;
	if (!(sp = sfstropen()))
		error(ERROR_SYSTEM|3, "out of space");
	sfputr(sp, usage_1, -1);
	if (error_info.id[2] == 'g')
		sfputr(sp, usage_grp_1, -1);
	else
	{
		sfputr(sp, usage_own_1, -1);
		options |= OPT_CHOWN;
	}
	sfputr(sp, usage_2, -1);
	if (options & OPT_CHOWN)
		sfputr(sp, ERROR_translate(0, 0, 0, "[owner[:group]]"), -1);
	else
		sfputr(sp, ERROR_translate(0, 0, 0, "[[owner:]group]"), -1);
	sfputr(sp, usage_3, -1);
	if (!(usage = sfstruse(sp)))
		error(ERROR_SYSTEM|3, "out of space");
	for (;;)
	{
		switch (optget(argv, usage))
		{
		case 'b':
			if (stat(opt_info.arg, &st))
				error(ERROR_exit(1), "%s: cannot stat", opt_info.arg);
			before = st.st_mtime;
			continue;
		case 'c':
		case 'v':
			options |= OPT_VERBOSE;
			continue;
		case 'f':
			options |= OPT_FORCE;
			continue;
		case 'h':
			options |= OPT_LCHOWN;
			continue;
		case 'm':
			memset(&mapdisc, 0, sizeof(mapdisc));
			mapdisc.key = offsetof(Map_t, key);
			mapdisc.size = sizeof(Key_t);
			if (!(map = dtopen(&mapdisc, Dtset)))
				error(ERROR_exit(1), "out of space [id map]");
			continue;
		case 'n':
			options |= OPT_SHOW;
			continue;
		case 'N':
			options |= OPT_NUMERIC;
			continue;
		case 'r':
			if (stat(opt_info.arg, &st))
				error(ERROR_exit(1), "%s: cannot stat", opt_info.arg);
			uid = st.st_uid;
			gid = st.st_gid;
			options |= OPT_UID|OPT_GID;
			continue;
		case 'u':
			options |= OPT_UNMAPPED;
			continue;
		case 'H':
			flags |= FTS_META|FTS_PHYSICAL;
			logical = 0;
			continue;
		case 'L':
			flags &= ~(FTS_META|FTS_PHYSICAL);
			logical = 0;
			continue;
		case 'P':
			flags &= ~FTS_META;
			flags |= FTS_PHYSICAL;
			logical = 0;
			continue;
		case 'R':
			flags &= ~FTS_TOP;
			logical = 0;
			continue;
		case 'X':
			options |= OPT_TEST;
			continue;
		case ':':
			error(2, "%s", opt_info.arg);
			continue;
		case '?':
			error(ERROR_usage(2), "%s", opt_info.arg);
			break;
		}
		break;
	}
	argv += opt_info.index;
	argc -= opt_info.index;
	if (error_info.errors || argc < 2)
		error(ERROR_usage(2), "%s", optusage(NiL));
	s = *argv;
	if (options & OPT_LCHOWN)
	{
		flags &= ~FTS_META;
		flags |= FTS_PHYSICAL;
		logical = 0;
	}
	if (logical)
		flags &= ~(FTS_META|FTS_PHYSICAL);
	if (map)
	{
		if (streq(s, "-"))
			sp = sfstdin;
		else if (!(sp = sfopen(NiL, s, "r")))
			error(ERROR_exit(1), "%s: cannot read", s);
		while (s = sfgetr(sp, '\n', 1))
		{
			getids(s, &t, &key, options);
			if (!(m = (Map_t*)dtmatch(map, &key)))
			{
				if (!(m = (Map_t*)stakalloc(sizeof(Map_t))))
					error(ERROR_exit(1), "out of space [id dictionary]");
				m->key = key;
				m->to.uid = m->to.gid = NOID;
				dtinsert(map, m);
			}
			getids(t, NiL, &m->to, options);
		}
		if (sp != sfstdin)
			sfclose(sp);
		keys[1].gid = keys[2].uid = NOID;
	}
	else if (!(options & (OPT_UID|OPT_GID)))
	{
		getids(s, NiL, &key, options);
		if ((uid = key.uid) != NOID)
			options |= OPT_UID;
		if ((gid = key.gid) != NOID)
			options |= OPT_GID;
	}
	switch (options & (OPT_UID|OPT_GID))
	{
	case OPT_UID:
		s = ERROR_translate(0, 0, 0, " owner");
		break;
	case OPT_GID:
		s = ERROR_translate(0, 0, 0, " group");
		break;
	case OPT_UID|OPT_GID:
		s = ERROR_translate(0, 0, 0, " owner and group");
		break;
	default:
		s = "";
		break;
	}
	if (!(fts = fts_open(argv + 1, flags, NiL)))
		error(ERROR_system(1), "%s: not found", argv[1]);
	while (!sh_checksig(context) && (ent = fts_read(fts)))
		switch (ent->fts_info)
		{
		case FTS_SL:
		case FTS_SLNONE:
			if (options & OPT_LCHOWN)
			{
#if _lib_lchown
				chownf = lchown;
				op = "lchown";
				goto commit;
#else
				if (!(options & OPT_FORCE))
				{
					errno = ENOSYS;
					error(ERROR_system(0), "%s: cannot change symlink owner/group", ent->fts_path);
				}
#endif
			}
			break;
		case FTS_F:
		case FTS_D:
		anyway:
			chownf = chown;
			op = "chown";
		commit:
			if ((unsigned long)ent->fts_statp->st_ctime >= before)
				break;
			if (map)
			{
				options &= ~(OPT_UID|OPT_GID);
				uid = gid = NOID;
				keys[0].uid = keys[1].uid = ent->fts_statp->st_uid;
				keys[0].gid = keys[2].gid = ent->fts_statp->st_gid;
				i = 0;
				do
				{
					if (m = (Map_t*)dtmatch(map, &keys[i]))
					{
						if (uid == NOID && m->to.uid != NOID)
						{
							uid = m->to.uid;
							options |= OPT_UID;
						}
						if (gid == NOID && m->to.gid != NOID)
						{
							gid = m->to.gid;
							options |= OPT_GID;
						}
					}
				} while (++i < elementsof(keys) && (uid == NOID || gid == NOID));
			}
			else
			{
				if (!(options & OPT_UID))
					uid = ent->fts_statp->st_uid;
				if (!(options & OPT_GID))
					gid = ent->fts_statp->st_gid;
			}
			if ((options & OPT_UNMAPPED) && (uid == NOID || gid == NOID))
			{
				if (uid == NOID && gid == NOID)
					error(ERROR_warn(0), "%s: uid and gid not mapped", ent->fts_path);
				else if (uid == NOID)
					error(ERROR_warn(0), "%s: uid not mapped", ent->fts_path);
				else
					error(ERROR_warn(0), "%s: gid not mapped", ent->fts_path);
			}
			if (uid != ent->fts_statp->st_uid && uid != NOID || gid != ent->fts_statp->st_gid && gid != NOID)
			{
				if (options & (OPT_SHOW|OPT_VERBOSE))
				{
					if (options & OPT_TEST)
					{
						ent->fts_statp->st_uid = 0;
						ent->fts_statp->st_gid = 0;
					}
					sfprintf(sfstdout, "%s uid:%05d->%05d gid:%05d->%05d %s\n", op, ent->fts_statp->st_uid, uid, ent->fts_statp->st_gid, gid, ent->fts_path);
				}
				if (!(options & OPT_SHOW) && (*chownf)(ent->fts_accpath, uid, gid) && !(options & OPT_FORCE))
					error(ERROR_system(0), "%s: cannot change%s", ent->fts_path, s);
			}
			break;
		case FTS_DC:
			if (!(options & OPT_FORCE))
				error(ERROR_warn(0), "%s: directory causes cycle", ent->fts_path);
			break;
		case FTS_DNR:
			if (!(options & OPT_FORCE))
				error(ERROR_system(0), "%s: cannot read directory", ent->fts_path);
			goto anyway;
		case FTS_DNX:
			if (!(options & OPT_FORCE))
				error(ERROR_system(0), "%s: cannot search directory", ent->fts_path);
			goto anyway;
		case FTS_NS:
			if (!(options & OPT_FORCE))
				error(ERROR_system(0), "%s: not found", ent->fts_path);
			break;
		}
	fts_close(fts);
	if (map)
		dtclose(map);
	return error_info.errors != 0;
}
